sdk/identity/azure-identity/src/managed_identity_source.cpp

```diff
@@ -449,16 +449,27 @@ std::unique_ptr<ManagedIdentitySource> ImdsManagedIdentitySource::Create(
       credName + " will be created" + WithSourceMessage("Azure Instance Metadata Service")
           + ".\nSuccessful creation does not guarantee further successful token retrieval.");
 
-  return std::unique_ptr<ManagedIdentitySource>(new ImdsManagedIdentitySource(clientId, options));
+  std::string imdsHost = "http://169.254.169.254";
+  std::string customImdsHost = Environment::GetVariable("AZURE_IMDS_CUSTOM_AUTHORITY_HOST");
+  if (!customImdsHost.empty())
+  {
+    IdentityLog::Write(
+        IdentityLog::Level::Informational, "Custom IMDS host is set to: " + customImdsHost);
+    imdsHost = customImdsHost;
+  }
+  Azure::Core::Url imdsUrl(imdsHost);
+  imdsUrl.AppendPath("/metadata/identity/oauth2/token");
+
+  return std::unique_ptr<ManagedIdentitySource>(
+      new ImdsManagedIdentitySource(clientId, imdsUrl, options));
 }
 
 ImdsManagedIdentitySource::ImdsManagedIdentitySource(
     std::string const& clientId,
+    Azure::Core::Url const& imdsUrl,
```

Comment:
BTW, the most efficient way should be to take it by value instead of a reference here, and then std::move it into the `m_request`'s constructor

Reply:
What are the factors that go into a decision for when it is more efficient to pass something by value vs by reference? Is it the size of the object (i.e. number of member fields), types of internal fields, or something else?

Reply:
Pass by value when you can use std::move instead of copy constructor
1. If you end up holding a copy `m_something`.
2. If all you do is forward that object to another one, which in its turn takes it by value and moves it.

In general, we want to use std::move instead of copy constructor whenever possible - uses the very same object, allocated once
Copy constructor: (taking string as example) can be as bad as allocating a buffer, copying all characters, destroying old string. When with std::move we simply transfer the ownership of that buffer.

We should not take object by value, and take by const ref instead of value if we do not end up holding a copy of it, i.e. we use the string only in constructor - get its size(), or append it to another string. Then const ref is more efficient and taking by value is wasteful.

Reply:
Fixed in https://github.com/Azure/azure-sdk-for-cpp/pull/5839